### PR TITLE
add avx512

### PR DESCRIPTION
**Pull request by gnzlbg**

### CODE REVIEW

**Comment by shepmaster** [comment]
Sigh. I had a manual from 2017, so no wonder I missed so much stuff.

File: src/lib.rs [hunk 1]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
-        // 16-17 - reserved
+        16 => avx512f,
+        17 => avx512dq,
         18 => rdseed,
         19 => adx,
         20 => smap,
-        // 21-24 - reserved
-        25 => intel_processor_trace
-        // 26-31 - reserved
+        21 => avx512ifma,
```
Comment (shepmaster) [inline src/lib.rs]: `avx512_ifma` to match the manual, please.

File: src/lib.rs [hunk 2]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
+        // 22-24 - TODO
```
Comment (shepmaster) [inline src/lib.rs]: Should add:

> Bit 22: Reserved.
> Bit 23: CLFLUSHOPT.
> Bit 24: CLWB.

File: src/lib.rs [hunk 3]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
+        25 => intel_processor_trace,
+        26 => avx512pf,
+        27 => avx512er,
```
Comment (shepmaster) [inline src/lib.rs]: I don't understand where you are getting these names:

> Bits 27 - 26: Reserved.

File: src/lib.rs [hunk 4]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
+        28 => avx512cd,
+        // 29 - TODO
```
Comment (shepmaster) [inline src/lib.rs]: > Bit 29: SHA. supports Intel® Secure Hash Algorithm Extensions (Intel® SHA Extensions) if 1.

File: src/lib.rs [hunk 5]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
+        30 => avx512bw,
+        31 => avx512vl
     });
 
     bit!(ecx, {
-        0 => prefetchwt1
+        0 => prefetchwt1,
+        1 => avx512vbmi,
```
Comment (shepmaster) [inline src/lib.rs]: `avx512_vbmi` to match the manual, please.

File: src/lib.rs [hunk 6]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
+        // 2-4 - TODO
```
Comment (shepmaster) [inline src/lib.rs]: > Bit 02: UMIP. Supports user-mode instruction prevention if 1.
> Bit 03: PKU. Supports protection keys for user-mode pages if 1.

File: src/lib.rs [hunk 7]
```diff
@@ -607,17 +608,46 @@ impl StructuredExtendedInformation {
...
+        // 5 - reserved
+        6 => avx512vbmi2,
+        // 7 - reserved
+        // 8-10 - TODO
+        11 => avx512vnni,
+        12 => avx512bitalg,
+        // 13 - reserved
+        14 => avx512vpopcntdq
```
Comment (shepmaster) [inline src/lib.rs]: Where are these coming from?

> Bits 16 - 5: Reserved.

File: src/lib.rs (hunk 2 above)
Comment (gnzlbg) [reply src/lib.rs]: I left everything non-AVX512 related out of this PR on purpose (basically everywhere where I wrote TODO). Should I complete those?

File: src/lib.rs (hunk 1 above)
Comment (gnzlbg) [reply src/lib.rs]: Note that, for whatever reason, the [Intel intrinsics guide](https://software.intel.com/sites/landingpage/IntrinsicsGuide) calls these `avx512ifma52`...while [The Manual (October 2017)](https://software.intel.com/sites/default/files/managed/39/c5/325462-sdm-vol-1-2abcd-3abcd.pdf) as you mention calls these `AVX512_IFMA`.

File: src/lib.rs (hunk 3 above)
Comment (gnzlbg) [reply src/lib.rs]: `avx512pf` and `avx512er` are in [the Intrinsics Guide]( https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=avx512er&expand=4104,3261,3264,3275,3278,3276,3273,3270,3267,3262,4104,2177,2176) and [The Manual (October 2017)](https://software.intel.com/sites/default/files/managed/39/c5/325462-sdm-vol-1-2abcd-3abcd.pdf).

File: src/lib.rs (hunk 5 above)
Comment (gnzlbg) [reply src/lib.rs]: Here the manual differs from the intrinsics guide again, `avx512vbmi` vs `avx512_vbmi`. I really don't understand when the Intel guys decide to use an `_` and when not. The intrinsics guide is more consistent than the manual.

File: src/lib.rs (hunk 7 above)
Comment (gnzlbg) [reply src/lib.rs]: The Linux kernel (this is what /proc/cpuinfo returns): https://patchwork.kernel.org/patch/10033543/

Note how the Intel devs themselves are unclear about the naming scheme, some are separated with an underscore, some are not. They might make the spec consistent at some point (the spec is allowed to change).

**Comment by gnzlbg** [comment]
I can slim this down to the AVX512 intrinsics implemented in the spec.

The CPUID for the features that re not in the spec yet won't change. There are already CPUs offering these in the wild, clang/gcc and the kernels support them already, etc. What might change before they get int the spec is the official CPUID name (e.g. `avx512vnni` vs `avx512_vnni`). I hope that for consistency with all the others they kill the underscores (as the Intel kernel devs argue) but we'll see.

**Comment by gnzlbg** [comment]
Ok reading the entire kernel discussion it seems that the mixed notation is going to stay in the spec and the kernel...  so the right names are `avx512vbmi` but `avx512_vbmi2`... sigh... I'll update this tomorrow to conform to the latest spec/kernel.

File: src/lib.rs (hunk 7 above)
Comment (gnzlbg) [reply src/lib.rs]: - `vpopcntdq`: https://software.intel.com/sites/landingpage/IntrinsicsGuide/#text=avx512vpop&expand=4104,3261,3264,3275,3278,3276,3273,3270,3267,3262,4104,2177,2176,3764,4082 

The others come from the kernel: https://patchwork.kernel.org/patch/10033543/

```c++
#define X86_FEATURE_AVX512_VBMI2 (16*32+ 6) /* Additional AVX512 Vector Bit Manipulation Instructions */
#define X86_FEATURE_GFNI	(16*32+ 8) /* Galois Field New Instructions */
#define X86_FEATURE_VAES	(16*32+ 9) /* Vector AES */
#define X86_FEATURE_VPCLMULQDQ	(16*32+ 10) /* Carry-Less Multiplication Double Quadword */
#define X86_FEATURE_AVX512_VNNI (16*32+ 11) /* Vector Neural Network Instructions */
#define X86_FEATURE_AVX512_BITALG (16*32+12) /* Support for VPOPCNT[B,W] and VPSHUF-BITQMB */
```

**Comment by gnzlbg** [comment]
So I've found the real reference for all these features, and updated the PR, from the commit msg:

> The reference is the Intel Architecture Instruction Set Extensions and Future
Features Programming Interface document (Tables 1-1. and 1-2.)
https://bugzilla.kernel.org/attachment.cgi?id=258813

**Comment by gnzlbg** [comment]
@shepmaster Rust has recently gained support for these :) 

Would it be possible to speed things up (and make a release afterwards) ? I hope I have addressed everything.

**Review by shepmaster — APPROVED** [review]

**Comment by shepmaster** [comment]
> Would it be possible to speed things up (and make a release afterwards)

Oops, I lost this in my other notifications — released 0.5.0!

**Comment by gnzlbg** [comment]
Awesome thanks!

**Pull request merged**